src/Std/Time/Zoned/ZonedDateTime.lean

```diff
@@ -0,0 +1,451 @@
+/-
+Copyright (c) 2024 Lean FRO, LLC. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sofia Rodrigues
+-/
+prelude
+import Std.Time.Zoned.DateTime
+import Std.Time.Zoned.ZoneRules
+
+namespace Std
+namespace Time
+
+set_option linter.all true
+
+/--
+Represents a date and time with timezone information.
+-/
+structure ZonedDateTime where
+  private mk::
+
+  /--
+  The plain datetime component, evaluated lazily.
+  -/
+  date : Thunk PlainDateTime
+
+  /--
+  The corresponding timestamp for the datetime.
+  -/
+  timestamp : Timestamp
+
+  /--
+  The timezone rules applied to this datetime.
+  -/
+  rules : TimeZone.ZoneRules
+
+  /--
+  The timezone associated with this datetime.
+  -/
+  timezone : TimeZone
+
+instance : Inhabited ZonedDateTime where
+  default := ⟨Thunk.mk Inhabited.default, Inhabited.default, Inhabited.default, Inhabited.default⟩
+
+namespace ZonedDateTime
+open DateTime
+
+/--
+Creates a new `ZonedDateTime` out of a `Timestamp` and a `TimeZone`.
+-/
+@[inline]
+def ofTimestamp (tm : Timestamp) (rules : TimeZone.ZoneRules) : ZonedDateTime :=
```

Comment:
There should be a variant of this function that takes a `TimeZone` instead of a `ZoneRules`, that just calls `ZoneRules.ofTimeZone`. Similar for `ofPlainDateTime` below.

There should also be a function that "fixes" a `ZonedDateTime` at the current time zone and forgets all other zone rules, to "opt out" of DST-aware datetime arithmetic. The Java equivalent is called [withFixedOffsetZone](https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/time/ZonedDateTime.html#withFixedOffsetZone()).

Reply:
I added the ` ofTimestampWithZone` function.